### DISCUSSION TITLE
如何拯救残废的HDR？

### DISCUSSION THREAD

**Discussion by AlphonseZetta10** [hooke007/mpv_PlayKit#345] [mpv 通用]
资源是《闪光的哈萨维》UHD BD-REMUX，屏幕是4K SDR
引用[一个Rip组的介绍](https://share.dmhy.org/topics/view/571844_Mobile_Suit_Gundam_Hathaway_01_BDrip_HEVC_4K.html)

> 4K Full-UHDBD-compatible
影片原生为SDR
HDR后调服务商是臭名昭著的Qtec
果不其然是直接映射的
打戏本身偏暗 哪怕是剧院观影视设备情况也有可能一团糟
不过细节还在可以看清水准

想知道能不能通过各种设置，使原盘资源也有好的观感

**Comment by natural-harmonia-gropius** [comment]
来点切片

ffmpeg -i input.mp4 -ss 00:00 -t 10 output.mp4

**Reply by AlphonseZetta10** [reply]
因为体积限制只切了开头7.5s
https://github.com/hooke007/MPV_lazy/assets/111181772/8dde1520-8ac6-431e-86a6-4dc1ffb487e2

**Comment by hooke007** [comment]
> vo_gpu

它的主要问题和紫罗兰剧场hdr一样的。亮度不足203
两种简单粗暴的解决方式：
1. 抬高亮度 --brightness
2. `--target-peak=100`

**Reply by AlphonseZetta10** [reply]
用V5版改了target-peak=100，观感确实好多了

**Reply by hooke007** [reply]
闪光的哈萨维没搜到相关的讨论，但是紫罗兰永恒花园的hdr调色用的参考白点是108nit，考虑到都是同一家做的后调所以很有可能流程是一样的。
追求一点更高精度的话 --target-peak=108 然后在此基础上可参考其它方式（比如 https://github.com/hooke007/MPV_lazy/discussions/272 ）进行下一步的调色

**Reply by AlphonseZetta10** [reply]
感谢
不过我对比了一下，发现target-peak=108时画面还是比sdr版暗一点
场景1
HDR REMUX👇 SDR Rip👇 Biibili👇
<img width=200 src="https://github.com/hooke007/MPV_lazy/assets/111181772/c715ea82-cce6-4251-a1ed-af51913e25dd"> <img width=200 src="https://github.com/hooke007/MPV_lazy/assets/111181772/7e4e937b-7d73-49f9-b362-8ce0dd4e28b9"> <img width="200" src="https://github.com/hooke007/MPV_lazy/assets/111181772/1b1cc06a-5041-47ac-bd24-31c500db593f">
场景2
<img width=200 src="https://github.com/hooke007/MPV_lazy/assets/111181772/d4dde111-952f-4c79-8a3e-7663e9d60fe0"> <img width=200 src="https://github.com/hooke007/MPV_lazy/assets/111181772/33944faa-d43d-4343-bbf9-675846c76e84"> <img width=200 src="https://github.com/hooke007/MPV_lazy/assets/111181772/dc0059bd-8e9d-4f71-9bfa-b2c735eb2420">

**Reply by natural-harmonia-gropius** [reply]
用这个 https://github.com/natural-harmonia-gropius/hdr-toys 然后 glsl-shader-opts=L_sdr=108 试试

**Reply by natural-harmonia-gropius** [reply]
另外 mpv 的标准白是常量 203， target-peak=108 只是把映射后的信号 * 203 / 108 这样子，结果非常不可控

**Reply by AlphonseZetta10** [reply]
不行，画面几乎都变成白色了，像吃了个闪一样

**Reply by natural-harmonia-gropius** [reply]
加 vo=gpu-next 了吗

**Reply by hooke007** [reply]
> 发现target-peak=108时画面还是比sdr版暗一点

符合预期，我说的这是简单粗暴的方法。

**Reply by AlphonseZetta10** [reply]
> 加 vo=gpu-next 了吗

加之后变成高反差黑白画了

**Reply by hooke007** [reply]
。。你确定你按照它的说明正确操作了吗
hdr-toys 原始设置
<img width=200 src="https://github.com/hooke007/MPV_lazy/assets/41094733/0c3b055f-ba78-4010-b4c7-8f47ad1cb8a4">

hdr-toys + command `set glsl-shader-opts L_sdr=108`
<img width=200 src="https://github.com/hooke007/MPV_lazy/assets/41094733/b57f934d-048a-40d1-ab12-b01ff7da6c0a">

>然后 glsl-shader-opts=L_sdr=108 试试

@natural-harmonia-gropius 你这参数目前在mpv.conf里不生效，只能命令生效

**Reply by AlphonseZetta10** [reply]
我把配置都删了之后重新试了一遍，这次画面正确了
`set glsl-shader-opts L_sdr=108`这个参数只能在命令行生效，有什么解决办法吗？

**Reply by hooke007** [reply]
命令，意思是可以用命令行，也可以用快捷键。
而且这个设置本来也不适合放在mpv.conf里

**Reply by AlphonseZetta10** [reply]
我的想法是把它放到一个与视频同名的.conf文件里，前面`target-peak=108`这个参数我就是这么设置的，这样子用起来比较方便

**Reply by natural-harmonia-gropius** [reply]
> 你这参数目前在mpv.conf里不生效，只能命令生效

bug，先有参数后插入着色器参数不会被替换。
https://github.com/mpv-player/mpv/issues/12039

> 快捷键

我自己绑按键用的，因为上面提到的问题放 conf 里生不生效我这边测下来有点看运气。

```
z                   set glsl-shader-opts L_sdr=100
x                   set glsl-shader-opts L_sdr=203
```

**Reply by hooke007** [reply]
按照原仓库的设计，`glsl-shader-opts` 在着色器挂载之前传递导致传空了对象所以无效。
既然你打算用 `--use-filedir-conf=yes` ，那就把相关参数都写进单文件设置里，不用原仓库的示例用法。

**Comment by hooke007** [comment]
我检查了 mpv-lazy 和hdr-toy的即插即用兼容性。需要做一些额外处理

具体参见 https://github.com/hooke007/MPV_lazy/discussions/272#discussioncomment-6686997

- 如果使用单文件设置（仅适用当前的案例）：
  1. 移除选项保存项目中的着色器条目（操作参考同前）
  2. 执行原仓库的说明（仅步骤1+2）
  3. 单文件设置中这样写
  ```ini
  vo=gpu-next

  target-prim=bt.2020
  target-trc=pq
  # 这里的着色器随版本自行改动
  glsl-shader=...
  glsl-shader=...
  glsl-shader=...

  glsl-shader-opts=L_sdr=108
  ```

**Reply by AlphonseZetta10** [reply]
再多问一句，`script-opts-append = save_global_props-props=volume`这一条加在单文件里似乎也是可行的对吧

**Reply by hooke007** [reply]
这是个全局设置为什么要加进单文件里。。。
这个脚本没有考虑过运行时的动态更改，也不符合设计预期

**Reply by AlphonseZetta10** [reply]
修改这一条的目的是防止“专用”shaders被应用到其他视频对吧。我加到单文件的目的是将两个项目解耦，免得下次更新的时候需要手动该配置。当然也是个临时的选择。

**Reply by hooke007** [reply]
> 修改这一条的目的是防止“专用”shaders被应用到其他视频对吧

不是。当前的设置下，单文件设置不会污染 glsl-shaders 的属性
相反，是防止驻留的 glsl-shaders 被应用到单文件中

**Reply by hooke007** [reply]
如果你不想修改保存中的条目(即跳过步骤 i)，也是可以的。

单文件设置中这样写
```ini
vo=gpu-next

target-prim=bt.2020
target-trc=pq
glsl-shaders-clr
glsl-shader=...

glsl-shader-opts=L_sdr=108
```

另外在播此单文件的时候也不建议运行时增加其它着色器，把你要加的着色器写在单文件设置里。
这样就不会影响原先记录 glsl-shaders 的属性值

```ini
vo=gpu-next

target-prim=bt.2020
target-trc=pq
glsl-shaders-clr
glsl-shader=...
glsl-shader=~~/shaders/xxxxx.glsl
glsl-shader=~~/shaders/yyyyy.glsl

glsl-shader-opts=L_sdr=108
```

**Reply by AlphonseZetta10** [reply]
嗯，刚刚是我考虑不周了

**Comment by AlphonseZetta10** [comment]
配置了单文件，目前体验非常完美
@hooke007 @natural-harmonia-gropius 感谢两位大佬的帮助

**Reply by hooke007** [reply]
保持此问题开启，作为一个实用案例

**Reply by AlphonseZetta10** [reply]
明白
